airbyte-cdk/bulk/toolkits/load-db/src/main/kotlin/io/airbyte/cdk/load/orchestration/db/legacy_typing_deduping/TableNameMapFactory.kt

```diff
@@ -32,24 +35,183 @@ class TableCatalogFactory(
+
+    /**
+     * Resolves table name collisions by adding a hash suffix when needed. Adds the processed name
+     * to the tracking set.
+     */
+    private fun resolveTableNameCollision(
+        originalName: TableName,
+        processedNames: MutableSet<String>
+    ): TableName {
+        val nameStr = originalName.toPrettyString()
+
+        return if (nameStr in processedNames) {
+            // Create a hash-suffixed name to avoid collision
+            val hash =
+                DigestUtils.sha1Hex("${originalName.namespace}&airbyte&${originalName.name}")
+                    .substring(0, 3)
+            val newName = "${originalName.name}_$hash"
+            TableName(originalName.namespace, newName).also {
+                processedNames.add(it.toPrettyString())
+            }
+        } else {
+            // Use original name and add to processed set
+            processedNames.add(nameStr)
+            originalName
+        }
+    }
+
+    /**
+     * Creates column name mapping with handling for potential collisions using incremental
+     * numbering, with advanced resolution for truncation cases.
+     */
+    private fun createColumnNameMapping(stream: DestinationStream): ColumnNameMapping {
+        val processedColumnNames = mutableSetOf<String>()
+        val columnMappings = mutableMapOf<String, String>()
+
+        stream.schema.asColumns().forEach { (columnName, _) ->
+            val originalColumnName =
+                finalTableColumnNameGenerator.getColumnName(columnName).displayName
+
+            // Get a unique column name by adding incremental numbers if necessary
+            val finalColumnName =
+                resolveColumnNameCollision(originalColumnName, processedColumnNames)
+
+            processedColumnNames.add(finalColumnName)
+            columnMappings[columnName] = finalColumnName
+        }
+
+        return ColumnNameMapping(columnMappings)
+    }
+
+    /**
+     * Resolves column name collisions by first trying incremental suffixes (_1, _2, etc.) If that
+     * doesn't work due to name truncation, uses the more powerful superResolveColumnCollisions.
+     *
+     * @param processedName The name after initial processing by the column name generator
+     * @param existingNames Set of names already used for other columns
+     */
+    private fun resolveColumnNameCollision(
+        processedName: String,
+        existingNames: Set<String>
+    ): String {
+        // If processed name is unique, use it
+        if (processedName !in existingNames) {
+            return processedName
+        }
+
+        // Try adding incremental suffixes until we find a non-colliding name
+        var counter = 1
+        var candidateName: String
+        var previousCandidate = processedName
+
+        do {
+            // Generate candidate name by adding numeric suffix
+            candidateName =
+                finalTableColumnNameGenerator.getColumnName("${processedName}_$counter").displayName
```

Comment:
nit: I think the old logic was approximately `getColumnName("$originalColumnName_$counter")` (whereas this is more like `getColumnName("${getColumnName(originalColumnName)}_$counter")`)

(... that might also be true for how the table names are being handled, now that I think about it)